File: vane-core/src/main/java/org/oddlama/vane/core/config/loot/LootDefinition.java
package org.oddlama.vane.core.config.loot;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.bukkit.NamespacedKey;
import org.bukkit.loot.LootTables;
import org.oddlama.vane.core.LootTable.LootTableEntry;
import org.oddlama.vane.util.ItemUtil;
import org.oddlama.vane.util.StorageUtil;

public class LootDefinition {

    private record Entry(double chance, int amountMin, int amountMax, String itemDefinition) {
        public Object serialize() {
            final HashMap<String, Object> dict = new HashMap<>();
            dict.put("Chance", chance);
            dict.put("AmountMin", amountMin);
            dict.put("AmountMax", amountMax);
            dict.put("Item", itemDefinition);
            return dict;
        }

        public static Entry deserialize(Map<String, Object> map) {
            if (!(map.get("Chance") instanceof Double chance)) {
                throw new IllegalArgumentException("Invalid loot table entry: chance must be a double!");
            }

            if (!(map.get("AmountMin") instanceof Integer amountMin)) {
                throw new IllegalArgumentException("Invalid loot table entry: amountMin must be a int!");
            }

            if (!(map.get("AmountMax") instanceof Integer amountMax)) {
                throw new IllegalArgumentException("Invalid loot table entry: amountMax must be a int!");
            }

            if (!(map.get("Item") instanceof String itemDefinition)) {
                throw new IllegalArgumentException("Invalid loot table entry: itemDefinition must be a String!");
            }

            return new Entry(chance, amountMin, amountMax, itemDefinition);
        }
    }

    public String name;
    public List<NamespacedKey> affectedTables = new ArrayList<>();
    private List<Entry> entries = new ArrayList<>();

    public LootDefinition(final String name) {
        this.name = name;
    }

    public String name() {
        return name;
    }

    public NamespacedKey key(final NamespacedKey baseKey) {
        return StorageUtil.namespacedKey(baseKey.namespace(), baseKey.value() + "." + name);
    }

    public LootDefinition in(final NamespacedKey table) {
        affectedTables.add(table);
        return this;
    }

    public LootDefinition in(final LootTables table) {
        return in(table.getKey());
    }

    private LootDefinition add(Entry entry) {
        entries.add(entry);
        return this;
    }

    public LootDefinition add(double chance, int amountMin, int amountMax, String itemDefinition) {
        return add(new Entry(chance, amountMin, amountMax, itemDefinition));
    }

    public Object serialize() {
        final HashMap<String, Object> dict = new HashMap<>();
        // Use capitalized keys for YAML generation: "Tables" and "Items"
        dict.put("Tables", affectedTables.stream().map(NamespacedKey::toString).toList());
        dict.put("Items", entries.stream().map(Entry::serialize).toList());
        return dict;
    }

    @SuppressWarnings("unchecked")
    public static LootDefinition deserialize(String name, Object rawDict) {
        if (!(rawDict instanceof Map<?, ?> dict)) {
            throw new IllegalArgumentException(
                "Invalid loot table: Argument must be a Map<String, Object>, but is " + rawDict.getClass() + "!"
            );
        }

        final var tableDict = (Map<String, Object>) dict;
        final Object tablesObj = tableDict.containsKey("Tables") ? tableDict.get("Tables") : tableDict.get("tables");
        if (!(tablesObj instanceof List<?> tables)) {
            throw new IllegalArgumentException(
                "Invalid loot table: 'tables' must be a list"
            );
        }

        final Object itemsObj = tableDict.containsKey("Items") ? tableDict.get("Items") : tableDict.get("items");
        if (!(itemsObj instanceof List<?> items)) {
            throw new IllegalArgumentException(
                "Invalid loot table: 'items' must be a list"
            );
        }

        final var table = new LootDefinition(name);
        for (final var e : tables) {
            if (e instanceof String key) {
                table.in(NamespacedKey.fromString(key));
            }
        }
        for (final var e : items) {
            if (e instanceof Map<?, ?> map) {
                table.add(Entry.deserialize((Map<String, Object>) map));
            }
        }

        return table;
    }

    public List<LootTableEntry> entries() {
        return entries
            .stream()
            .map(e ->
                new LootTableEntry(
                    e.chance,
                    ItemUtil.itemstackFromString(e.itemDefinition).getLeft(),
                    e.amountMin,
                    e.amountMax
                )
            )
            .toList();
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/loot/LootDefinition.kt
package org.oddlama.vane.core.config.loot

import org.bukkit.NamespacedKey
import org.bukkit.loot.LootTables
import org.oddlama.vane.core.LootTable
import org.oddlama.vane.util.ItemUtil
import org.oddlama.vane.util.StorageUtil.namespacedKey

/**
 * Config definition for additional entries injected into vanilla loot tables.
 *
 * @param name logical loot-definition name.
 */
class LootDefinition(val name: String?) {
    /**
     * Internal serializable loot entry representation.
     */
    private data class Entry(
        /** Drop chance in percent-like units used by vane loot table processing. */
        val chance: Double,
        /** Minimum amount for generated stack size. */
        val amountMin: Int,
        /** Maximum amount for generated stack size. */
        val amountMax: Int,
        /** Item definition string parsed via [ItemUtil.itemstackFromString]. */
        val itemDefinition: String?
    ) {
        /** Serializes this entry to dictionary form. */
        fun serialize(): Map<String, Any?> = mapOf(
            "Chance" to chance,
            "AmountMin" to amountMin,
            "AmountMax" to amountMax,
            "Item" to itemDefinition
        )

        /**
         * Entry deserialization helpers.
         */
        companion object {
            /** Deserializes an [Entry] from map data. */
            fun deserialize(map: Map<String?, Any?>): Entry {
                val chance = when (val v = map["Chance"]) {
                    is Double -> v
                    is Float  -> v.toDouble()
                    is Int    -> v.toDouble()
                    is Long   -> v.toDouble()
                    else -> throw IllegalArgumentException("Invalid loot table entry: chance must be a number (double)!")
                }
                val amountMin = when (val v = map["AmountMin"]) {
                    is Int    -> v
                    is Double -> v.toInt()
                    is Long   -> v.toInt()
                    else -> throw IllegalArgumentException("Invalid loot table entry: amountMin must be an int!")
                }
                val amountMax = when (val v = map["AmountMax"]) {
                    is Int    -> v
                    is Double -> v.toInt()
                    is Long   -> v.toInt()
                    else -> throw IllegalArgumentException("Invalid loot table entry: amountMax must be an int!")
                }
                val itemDefinition = map["Item"]?.toString()
                return Entry(chance, amountMin, amountMax, itemDefinition)
            }
        }
    }

    /** Loot tables affected by this definition. */
    val affectedTables: MutableList<NamespacedKey?> = mutableListOf()

    /** Serialized loot entries belonging to this definition. */
    private val entries: MutableList<Entry> = mutableListOf()

    /** Builds a unique key for this definition under a base key. */
    fun key(baseKey: NamespacedKey): NamespacedKey =
        namespacedKey(baseKey.namespace(), "${baseKey.value()}.$name")

    /** Adds an affected loot table by namespaced key. */
    fun `in`(table: NamespacedKey?): LootDefinition = apply { affectedTables.add(table) }

    /** Adds an affected loot table from Bukkit [LootTables]. */
    fun `in`(table: LootTables): LootDefinition = `in`(table.key)

    /** Adds a prebuilt entry. */
    private fun add(entry: Entry): LootDefinition = apply { entries.add(entry) }

    /** Adds a loot entry from primitives and item definition string. */
    fun add(chance: Double, amountMin: Int, amountMax: Int, itemDefinition: String?): LootDefinition =
        add(Entry(chance, amountMin, amountMax, itemDefinition))

    /** Serializes this loot definition to dictionary form. */
    fun serialize(): Map<String, Any> = mapOf(
        "Tables" to affectedTables.map { it.toString() },
        "Items"  to entries.map { it.serialize() }
    )

    /** Converts serialized entries into runtime loot-table entries. */
    fun entries(): List<LootTable.LootTableEntry> = entries.map { e ->
        LootTable.LootTableEntry(
            e.chance,
            ItemUtil.itemstackFromString(e.itemDefinition!!).getLeft()!!,
            e.amountMin,
            e.amountMax
        )
    }

    /**
     * Deserialization helpers.
     */
    companion object {
        /** Deserializes a [LootDefinition] from dictionary form. */
        fun deserialize(name: String?, rawDict: Any): LootDefinition {
            require(rawDict is Map<*, *>) {
                "Invalid loot table: Argument must be a Map<String, Object>, but is ${rawDict.javaClass}!"
            }

            val tableDict = rawDict.entries
                .filter { it.key is String }
                .associate { it.key as String to it.value }

            val tablesObj = tableDict["Tables"] ?: tableDict["tables"]
            require(tablesObj is List<*>) { "Invalid loot table: 'tables' must be a list" }

            val itemsObj = tableDict["Items"] ?: tableDict["items"]
            require(itemsObj is List<*>) { "Invalid loot table: 'items' must be a list" }

            return LootDefinition(name).also { table ->
                tablesObj.filterIsInstance<String>().forEach { table.`in`(NamespacedKey.fromString(it)) }
                itemsObj.filterIsInstance<Map<*, *>>().forEach { e ->
                    val mapped = e.entries
                        .filter { it.key is String }
                        .associate { it.key as String? to it.value }
                    table.add(Entry.deserialize(mapped))
                }
            }
        }
    }
}
